### ISSUE TITLE
Performance investigation & results

### ISSUE THREAD

**Issue by alubbe** [jadejs/jade#1975]
I've been wanting to give an update on how my benchmarks are coming along and what I have been finding. My findings are about increasing the number of HTML strings per second generated by a compiled, cached function - not about improving the speed of lexing or parsing.

My approach is to analyse and improve the output produced by these jade files: https://github.com/alubbe/jade-ejs-bench/tree/master/jade
I  am taking the output produced by jade and comparing it with a) output produced by EJS and b) best practices for letting v8 optimize functions.

Here are the current findings:
- [x] jade's escape function is slower than EJS' (20-30%, fixed with https://github.com/jadejs/jade/pull/1976)
- [x] Array.join is slower than +=, especially for a small number of items (15-50%, fixed with https://github.com/jadejs/jade/pull/1977)
- [x] jade's ```each``` has to perform runtime checks whether to iterate of an array or object, making it a bit slower than two separate iteration implementations (<5%, let's leave it as is)
- [x] jade's ```each``` runs in an IIFE for variable scoping, losing quite a bit of performance (~20%, fixed by https://github.com/jadejs/jade/pull/1983)
- [x] jade's mixin implementation is really, really slow. It runs about 400% slower than EJS' functions. (fixed by https://github.com/jadejs/jade/pull/1982 - turned out to be a v8 bug)

I'm continuing my search in irregular intervals and will report them here. Would love to hear feedback and/or further things to investigate.
Sources:
https://github.com/alubbe/jade-ejs-bench
http://jsperf.com/jade-better-escape
http://jsperf.com/html-compilers/2
http://jsperf.com/html-compilers/3

**Comment by alubbe** [comment]
Updated to reflect the fixes of points 4 & 5.
Fixing point 3 doesn't really yield a lot of performance and would be a breaking change, so I have dropped pursuing it.

Here are the final results of https://github.com/alubbe/jade-ejs-bench

*alubbe's allimprovs branch*
mixin-escape: 7457ms
mixin-no-escape: 2305ms
no-mixin-escape: 7884ms
no-mixin-no-escape: 1313ms
simple: 259ms

*jade 1.10.0*
mixin-escape: 44930ms
mixin-no-escape: 20945ms
no-mixin-escape: 28157ms
no-mixin-no-escape: 6409ms
simple: 404ms

*EJS 2.3.1*
mixin-escape: 6608ms
mixin-no-escape: 1653ms
no-mixin-escape: 7710ms
no-mixin-no-escape: 1456ms
simple: 310ms

*EJS-TJ 2.3.2*
mixin-escape: 5817ms
mixin-no-escape: 4074ms
no-mixin-escape: 7194ms
no-mixin-no-escape: 4063ms
simple: 183ms

I am very happy with those numbers!
@TimothyGu I'd love for you to bring the ```+=``` to EJS, as well. It should speed things up for that project, as well.
